extensions/amp-date-picker/0.1/amp-date-picker.js

```diff
@@ -210,85 +309,572 @@ class AmpDatePicker extends AMP.BaseElement {
     /** @private {?function(?):?} */
     this.templateThen_ = null;
 
+    /** @private {?Element} */
+    this.startDateField_ = null;
+
+    /** @private {?Element} */
+    this.endDateField_ = null;
+
+    /** @private {?Element} */
+    this.dateField_ = null;
+
+    /** @private */
+    this.renderedTemplates_ = map();
+
+    /** @private {?function()} */
+    this.templatesReadyResolver_ = null;
+
+    /** @private {!Promise} */
+    this.templatesReadyPromise_ = new Promise(resolve => {
+      this.templatesReadyResolver_ = resolve;
+    });
+
+    /** @private @const {!Array<!UnlistenDef>} */
+    this.unlisteners_ = [];
+
+    const initialState = (this.mode_ == DatePickerMode.OVERLAY ?
+      DatePickerState.OVERLAY_CLOSED :
+      DatePickerState.STATIC);
     /** @private @const */
-    this.registerAction_ = this.registerAction.bind(this);
+    this.stateMachine_ = new FiniteStateMachine(initialState);
+    this.setupStateMachine_(this.stateMachine_);
 
     const locale = this.element.getAttribute('locale') || DEFAULT_LOCALE;
     this.moment_.locale(locale);
   }
 
   /** @override */
   isLayoutSupported(layout) {
-    return layout == Layout.CONTAINER;
+    return layout == Layout.CONTAINER || layout == Layout.NODISPLAY;
+  }
+
+  /** @override */
+  layoutCallback() {
+    this.setupTemplates_();
+    // Make sure it's rendered and measured properly
+    return this.render(this.state_);
+  }
+
+  /** @override */
+  resumeCallback() {
+    this.setupListeners_();
+    this.setupTemplates_();
+  }
+
+  /**
+   * TODO(cvializ): Figure out how to keep the listeners when unlayout callback
+   * is triggered by a parent element e.g. amp-lightbox collapsing.
+   * @override
+   */
+  unlayoutCallback() {
+    this.cleanupListeners_();
+    this.cleanupSrcTemplates_();
+    this.clearRenderedTemplates_();
+    return true;
+  }
+
+  /** @override */
+  unlayoutOnPause() {
+    return true;
```

Comment:
if this is `false` which I believe should be, I can't see anyway that closing a lightbox could unlayout this element. Could you try it and see if it still gets unlayed out when lightbox closes?

Reply:
I found the `detachedCallback` is a better method for deleting the listeners, so I'll use that instead of `unlayoutCallback` and `pauseCallback`.